[#142][FEATURE] 그룹 해시태그 기능 추가

unikal1 opened the pull request:
<!-- PR의 제목은 "[#이슈번호] [종류(ex.FEATURE)] 이슈이름" 으로 작성 -->
테스트 개편에 대한 부분이 많습니다.  실제 비지니스 로직 부분은 그리 많지 않습니다. 
<!-- Auto close 할 이슈 번호 -->
- close #142

## ✨ 구현 기능 명세

## 그룹 hashTag  create
각 그룹 생성 시 hashTag 를 함께 저장할 수 있습니다. 해시태그의 유무는 선택이며, 최대 개수 5개, 각 문자열의 길이는 20 입니다.

이는 `HashTag` 엔티티와 `GroupHashTag` 엔티티를 통해 관리합니다. `HashTag` 엔티티는 정규화된 문자열을 기준으로 하여 실제 데이터를 저장하고 있으며 추후 검색을 위해 정규화된 값입니다. 사용자가 해시태그에 대해 검색 시 해당 엔티티를 거칩니다.

`groupHashTag` 엔티티는 groupId, hashTagId 뿐만 아니라 실제 hashtag 의 원본을 가지고 있습니다. 이는 실제로 그룹 검색 시 클라이언트에게 보여지는 부분이며, 정규화되기 이전의 값입니다. 대부분, 사용자가 입력한 데이터 원본입니다.


현재는 group create 시에만 hashTag 가 같이 저장하는 부분만 구현되어 있습니다. 

## RetryTx
`@Transactional` 어노테이션의 대체제이며, 실패 시 새로운 트랜잭션을 열어 재시도하는 어노테이션입니다. 자세한 내용은 [retryTx 재시도 전략](https://github.com/study-pals/backend/wiki/retry-%EB%A1%9C%EC%A7%81) 를 참고해주세요.

## test 및 build 환경 개편

`run/debug configurations` 에서 VM options 에 다음 두 값을 입력할 수 있습니다.

`-Dspring.profiles.active`, `-Dtc.disable` 이며, 각각 `local, test`, `true, false` 가 올 수 있습니다.

전자는 테스트 및 빌드 시 어떤 프로필을 사용할지에 대한 내용이며, 기본 default 에 덮어씌워질 수 있습니다.

후자는 `TestContainer` 활성 여부를 선택하며, false 선택 시 활성화된 profile 로 테스트를 수행할 수 있습니다.

`local`, `false` 조합 사용 시, localhost 로 빌드/테스트를 수행할 수 있습니다.

unikal1 committed:
Feat: 해시태그 엔티티 추가

- hashTag 및 group 과의 매핑 테이블 groupHashTag 정의
- mapping table 상에서 group/hashtag 에 대한 lazy loading
- tag 에 대한 유니크(+인덱싱)
- groupHashTag 에서 group 에 대한 인덱싱

Ref: #132

unikal1 committed:
Feat: 해시태그 dao 추가

- usedCount 필드 원자적 증감/soft delete 추가
- n+1 방지 fetch join 메서드 추가
- 자동완성 용 prefix 용 조회 메서드 추가

Ref: #132

unikal1 committed:
Feat: displayTag 추가

- 실제 저장되는 값과, 보여지는 데이터를 분리하여 보여지는 데이터를 groupHashTag 에 추가

Ref: #132

unikal1 committed:
Feat: 그룹 생성 시 hash tag 추가 로직 추가(service~)

- 해시태그 정규화 및 ui 상 보여지는 이름 분리(실제 저장과 외부 노출 값 분리)
- 동시성 문제로 인한 unique 예외에 대비한 1회 재시도 전략
- 이미 존재하는 태그는 1 추가/ 아닌 경우는 새롭게 생성

Ref: #132

unikal1 committed:
Feat: groupHashTag 에 group-hashtag unique 제약 조건 추가

Ref: #132

unikal1 commented on src/main/java/com/studypals/domain/groupManage/service/GroupServiceImpl.java:
```diff
@@ -60,6 +62,9 @@ public Long createGroup(Long userId, CreateGroupReq dto) {
         groupMemberWriter.createLeader(member, group);
 
+        // 해시태그 삽입 (12-23 #132 sang)
+        groupHashTagWorker.saveTags(group, dto.hashTags());
```
groupHashTagWorker 에서 null 체크를 하는게 더 좋지 않을까?

unikal1 committed:
Fix : copilot 리뷰 적용

- naming, null check, 일부 잠재적인 동시성 오류 부분 수정

Ref: #142

sleepyhoon commented on src/main/java/com/studypals/domain/groupManage/worker/GroupHashTagWorker.java:
```diff
@@ -0,0 +1,157 @@
+    private Map<String, String> toNormalizedAndRowMap(List<String> tags) {
+        Map<String, String> result = new HashMap<>();
+        for (String tag : tags) {
+            if (tag.isEmpty()) continue;
```
[P3]
`CreateGroupReq`에서 tags를 받을 때 @isBlank 검사를 진행하고 있어서 isEmpty 검사를 안해도 되지 않을까요? 안전하게 추가 검사를 진행하는 용도라면 그대로 진행하면 될 것 같습니다.

unikal1 replied:
추후 hashtag 관련하여 기능이 확장된다면 재사용을 염두해 두고 있긴 합니다. 현재의 로직 상에는 말씀하신 것 처럼 필요 없어 보이긴 하는데 코드를 재사용하다 NPE 위험이 있기에 (오버헤드도 적고) 현재 방식으로 가고자 합니다.

sleepyhoon commented on src/main/java/com/studypals/domain/groupManage/worker/GroupHashTagWorker.java:
```diff
@@ -0,0 +1,157 @@
+        // 저장된 hashtag 를 기반으로 groupHashTag 저장
+        List<GroupHashTag> groupHashTags = new ArrayList<>();
+        for (HashTag tag : exists) {
+            String val = normalized.getOrDefault(tag.getTag(), tag.getTag());
```
[P2]
`normalized.getOrDefault(tag.getTag(), tag.getTag())`에서 좌우가 모두 `tag.getTag()`인데 의도하신 부분인가요?

unikal1 replied:
네. getOrDefault 의 첫번째 파라미터는 map 에서 검색할 key 값, 2번째는 없으면 제공되는 값입니다. 즉, getTag 기반으로 찾되, 없으면 해당 값을 그대로 사용한다는 취지입니다.

sleepyhoon commented on src/main/java/com/studypals/domain/groupManage/worker/GroupHashTagWorker.java:
```diff
@@ -0,0 +1,157 @@
+                hashTagRepository.saveAll(notExists);
+                hashTagRepository.flush(); // unique 제약 조건 발생용 flush
+            } catch (DataIntegrityViolationException e) { // 저장 실패 시 동시성 문제라 생각하고 1회 재시도
+                retrySave(notExists.stream().map(HashTag::getTag).toList());
+            }
+        }
```
[P1]
여기서 예외가 발생하면 서비스단에서 만든 트랜잭션이 모두 롤백이 되기 때문에 추가 로직을 고려해야 할 수 있을 것 같습니다.

sleepyhoon replied:
동시성 문제 관련해서 낙관적 락을 말했었는데, 말씀 주신대로 새로 insert 되는 데이터는 낙관적 락이 안되는 것으로 보입니다.

그래서 일단은 native query를 사용하여 MySQL 에서 사용하는 `INSERT IGNORE` 을 사용해서 구현하는게 가장 깔끔할 것 같습니다.

unikal1 replied:
retryTx 어노테이션을 통해 트랜잭션 자체를 재시도 하도록 하였습니다. 자세한 내용은 위키를 참고해 주세요.

sleepyhoon replied:
확인했습니다. 주석이 부분은 아직 미완료된거 같아 완료 후 resolve 부탁드립니다.

sleepyhoon commented on src/main/java/com/studypals/domain/groupManage/entity/HashTag.java:
```diff
@@ -0,0 +1,40 @@
+    @Column(name = "tag", nullable = false, unique = true)
+    private String tag;
+
+    @Builder.Default
+    @Column(name = "used_count", nullable = false)
+    private Long usedCount = 1L;
```
[P2]
이 값의 의도가 궁금합니다.

unikal1 replied:
해당 해시태그를 사용하고 있는 group 의 개수를 파악합니다. 이는, hashTag 의 안전한 삭제를 위함입니다.

어떠한 그룹도 사용하지 않는 해시태그의 경우 리소스 절약을 위해 삭제해야 됩니다. 그러나, 매번 groupHashTag 에서 count 를 통해 개수를 세는 것은 오버헤드가 크다고 판단되어 그룹이 해당 해시태그를 참조하고 있으면 1 증가 / 그룹 해시태그 변경 , 해시태그 삭제, 그룹 삭제 시 1 감소시켜 0이 되면 추후 배치 서버 등을 통해 지우게끔 하고자 합니다.

sleepyhoon commented on src/main/java/com/studypals/domain/groupManage/dao/HashTagRepository.java:
```diff
@@ -0,0 +1,106 @@
+     * @param tag 감소시킬 태그
+     * @return 변경된 row 수
+     */
+    @Modifying(clearAutomatically = true)
```
[P2]
(clearAutomatically = true) 을 사용해야 하는지 궁금합니다.

unikal1 replied:
1차 캐시를 비웁니다. 이는, modifying 된 결과 값이 1차 캐시에 적용되지 않기 때문에, 해당 쿼리문 실행 이후 1차 캐시를 비운 뒤 해당 데이터를 재요청하는 경우 db에서 검색해 옵니다. 
다만, 처음 생각할 때는 1 감소 이후 usedCount 가 0 이면 바로 삭제하게끔 하기 위해 `decreaseUsedCount` 이후 `usedCount` 검사 후 `hashTagRepository.delete()` 를 하고자 하여 해당 clearAutomatically 를 적용하였으나 중간에 관리 책임을 배치 서버로 돌리고자 하였고 현재로선 필요가 없어 보입니다. 

해당 usedCount 값을 변경하고 이를 다시 읽어온다면 해당 옵션이 필요하나, increase 관련한 메서드의 경우, 1차 캐시를 지우면 꽤나 많은 데이터를 재조회해야 하기도 하고, 변경 후 변경된 데이터를 사용하지 않으므로 지우는 것으로 하겠습니다.

sleepyhoon replied:
확인했습니다. 해당 코드를 지우고 resolve 해주세요.

sleepyhoon reviewed (commented):
1차 리뷰입니다.

unikal1 committed:
Feat: 재시도용 retry 어노테이션 작성

- transactional 을 덮는 RetryTx 어노테이션 추가

Ref: #142

unikal1 committed:
Test: retryTx aspect 에 대한 테스트

- retry 정상 작동 여부 테스트
- hashTag 관련한 변경으로 인해 실패하는 테스트 수정
- aspect 시 annotation 이 제대로 들어가지 않는 문제 해결
- 주석 추가

Ref: #142

unikal1 committed:
Refactor: RetryTx 적용 후, 자체적인 재시도 전략 제거

- try-catch 제거
- retryTx 적용

Ref: #142

unikal1 committed:
Fix : clearAutomatically 제거

- hashtag 삭제를 batch 서버로 이전함에 따라, 1차 캐시 초기화 불필요로 인한 삭제

Ref: #142

unikal1 committed:
Fix : findByPrefix -> 일반적인 검색 명령어 search 로 변경

- 기존 prefix 를 토대로 검색을 대신하여 중간에 값이 있어도 검색 가능하도록 변경

Ref: #142

unikal1 committed:
Build: application.properties 정리 및 테스트 환경 개편

- local/ test / prod 분리 및 운용 DB 분리
- 자세한 내용은 위키 참조

Ref: #142

unikal1 committed:
Test: 테스트 반영

- test 프로파일 적용 제거
- hash tag 테스트 추가
- retryTx 성능 테스트 시 db 초기화 로직 추가
- 문자열 정규화 로직 추가
- 그 외 테스트 유틸 수정

Ref: #142

unikal1 committed:
Build: 테스트 시 활성화되는 profile override 설정

- 기존에 -Dspring.profiles.active 환경변수가 있으면 이를 사용, 없으면 test 프로필 적용

Ref: #142

unikal1 committed:
Build: 환경변수 인자 수정

- 잘못 작성된 이름 수정

Ref: #142

unikal1 commented:
@sleepyhoon 리뷰 반영 및 retryTx 어노테이션 관련 추가, 일부 자잘한 로직 등을 개편하였습니다.

application.properties 구조를 개편하였습니다. 중복된 부분을 제거하고 각 상황에 맞게 prod(application.properties), local(application-local.properties), test(application-test.properties)로 구분하였습니다.

`tc.disable` vm 환경변수를 사용하여 test container 사용 여부를 분리하였습니다. 빌드/테스트 시 해당 변수를 이용해 임시 컨테이너를 사용하여 테스트할지 등을 볼 수 있습니다.

`spring.profiles.active` 를 덮어씌울 수 있습니다. vm 환경변수를 통해 test 시 test 프로필 외에도 다른 프로필을 임의로 적용할 수 있습니다.

sleepyhoon commented on src/main/java/com/studypals/domain/groupManage/dao/HashTagRepository.java:
```diff
@@ -0,0 +1,106 @@
+     */
+    Optional<HashTag> findByTag(String tag);
+
+    /**
+     * tag 자동완성 시 사용할 메서드. prefix 에 대해 cnt 값 만큼의 자주 사용되는 데이터를 반환합니다.
+     *
+     * @param value 검색할 인자(접두사 / 순서대로)
+     * @param pageable 반환 개수 지정
+     * @return cnt 개수 만큼의, 사용 빈도가 높은 데이터
+     */
+    @Query(
+            """
+    SELECT t.tag
+    FROM HashTag t
+    WHERE t.tag LIKE CONCAT('%', :value, '%')
+    ORDER BY t.usedCount DESC
+""")
+    List<String> search(@Param("value") String value, Pageable pageable);
```
[P2]
접두사로 찾는게 아니라 `CONCAT('%', :value, '%')` 으로 작성하면 인덱스를 이용하지 못하는데, 성능 상 문제가 있을 것 같습니다. 이전의 방식이 낫지 않을까요?

unikal1 replied:
해당 문제는 인지하고 있습니다. 다만 검색이라는 기능의 취지를 생각해보면 앞 부분이 일치하는지 여부만 확인하는 과정이 너무 좁은 범위지 않나  싶습니다.
아마도 추후에는 elastic search를 사용하지 않을까요?

sleepyhoon replied:
확인했습니다.

sleepyhoon commented on build.gradle:
```diff
@@ -139,7 +139,12 @@ tasks.named('test', Test) {
     outputs.dir(snippetsDir)
     environment envProps
-    systemProperty 'spring.profiles.active', 'test'
+    def active =
+            System.getProperty("spring.profiles.active")
+                    ?: System.getenv("SPRING_PROFILES_ACTIVE")
+                    ?: "test"
+
+    systemProperty "spring.profiles.active", active
```
[P3]
test를 돌릴 때는 항상 spring.profiles.active 값을 test로 하는게 자연스럽지 않나요? 해당 코드가 추가된 이유가 궁금합니다.

unikal1 replied:
그래서 default 값은 test 가 맞습니다. 다만 각 로컬에서 커스터마이징이 가능하도록 확장한 것입니다.
해당 코드는 profile 을 test 로 두되, 만약 임의로 환경변수를 추가로 설정한다면 해당 프로필을 사용하도록 둔 것입니다.

sleepyhoon commented on src/main/java/com/studypals/domain/groupManage/dao/HashTagRepository.java:
```diff
@@ -0,0 +1,106 @@
+    /**
+     * tag 자동완성 시 사용할 메서드. prefix 에 대해 cnt 값 만큼의 자주 사용되는 데이터를 반환합니다.
+     *
+     * @param value 검색할 인자(접두사 / 순서대로)
```
[P3]
만약 접두사 형식으로 하지 않을 경우 해당 주석을 수정해야 할 것 같습니다.

unikal1 replied:
확인했습니다.

sleepyhoon reviewed (commented):
2차 리뷰 입니다

sleepyhoon commented on src/main/resources/application-local.properties:
```diff
@@ -1,39 +1,26 @@
+# ===============================
+spring.datasource.url=jdbc:mysql://127.0.0.1:3306/study_pal?useSSL=false&allowPublicKeyRetrieval=true&useUnicode=true&serverTimezone=Asia/Seoul
+spring.datasource.username=root
+spring.datasource.password=root
```
[P1]
로컬이기는 하지만 환경변수 처리하는게 안전할 것 같습니다.

sleepyhoon commented on src/main/resources/application-local.properties:
```diff
@@ -1,39 +1,26 @@
+# Redis (local)
+# ===============================
+spring.data.redis.host=localhost
+spring.data.redis.port=6379
```
[P1]
로컬이기는 하지만 환경변수 처리하는게 안전할 것 같습니다.

unikal1 replied:
환경변수 처리를 안하기 위해서 local 을 따로 둔 것입니다. 차라리 .gitIgnore 에 application-local.properties 를 둘까요?

sleepyhoon replied:
그게 나을 것 같습니다.

sleepyhoon reviewed (commented):
properties 파일 리뷰입니다.

unikal1 committed:
Fix: PR 리뷰 적용

- search 관련 주석 변경
- application-local.properties 가 더 이 상 git 에서 추적되지 않음.

Ref: #142

sleepyhoon commented:
LGTM

unikal1 commented:
@codex review

unikal1 commented on src/main/java/com/studypals/global/retry/RetryTxAspect.java:
```diff
@@ -0,0 +1,172 @@
+    @Around("@annotation(com.studypals.global.retry.RetryTx)")
+    public Object around(ProceedingJoinPoint pjp) throws Throwable {
+        // 어노테이션 설정값 조회
+        MethodSignature methodSignature = (MethodSignature) pjp.getSignature();
+        RetryTx retryTx = methodSignature.getMethod().getAnnotation(RetryTx.class);
+
+        // 방어적 보정 (이상값을 넣어도 최소 1회는 실행되도록)
+        int maxAttempts = Math.max(1, retryTx.maxAttempts());
```
@codex 
그러나, 해당 어노테이션을 interface 가 아닌 구현 메서드에만 사용한다면?

unikal1 replied:
spring aop 는 CGLIB 를 default 로 사용하고 있으며 이 경우 해당 문제는 일어나지 않습니다. resolve.

unikal1 committed:
Fix: 기본 해시태그 null 입력을 빈 리스트로 처리

Ref: #158

unikal1 committed:
Merge branch 'dev' into feat/group-hashtag
